woocommerce/woocommerce-ios
# Variable Products are now editable with readonly variations

Fixes #2084 

- [x] ⚠️ Make sure PR base is `develop` after https://github.com/woocommerce/woocommerce-ios/pull/2508 is merged ⚠️ 

## Changes

- Added `variationsImage` with `types` icon from Gridicons for the variations row in product form + unit test
- Added `variations` case to `ProductFormEditAction` enum for the variations row. In `ProductFormActionsFactory`, it only shows the variations row for a variable product if it has non-zero variations
- Added a helper `isNotEmpty` to `ProductFormSection` enum, so that we can filter the visible sections in `DefaultProductFormTableViewModel`
- In `DefaultProductFormTableViewModel`, configured the variations row view model
- In `ProductFormViewController`, showed the readonly variations when the user taps on the variations row in product form
- In `ProductDetailsFactory`, updated to show editable product form instead of readonly product details for a variable product if M3 flag is enabled
- Updated unit tests

## Testing

Prerequisites: the site has at least 2 variable products, one has no variations and the other has non-zero variations

### M3 feature flag is on: variable products are editable

**Test case 1**

- Go to the Products tab
- Tap on a variable product without variations --> should not see a variations row
- Feel free to edit the product name, description, images, tags, categories, short description
- Tap "Update" --> the product should be updated remotely

**Test case 2**

- Go to the Products tab
- Tap on a variable product with variations --> should see a variations row
- Tap the variations row --> it should show readonly variations
- Go back to the product form, and free to edit the product name, description, images, tags, categories, short description
- Tap "Update" --> the product should be updated remotely

### M3 feature flag is on: variable products are readonly

Please turn off M3 feature flag by returning `false` for `editProductsRelease3` in `DefaultFeatureFlagService`

**Test case 1**

- Go to the Products tab
- Tap on a variable product without variations --> it should show the readonly product details

**Test case 2**

- Go to the Products tab
- Tap on a variable product with variations --> it should show the readonly product details

## Example screenshots

![simulator](https://user-images.githubusercontent.com/1945542/87510371-0b9e0e80-c6a6-11ea-8ff4-a40367949c13.gif)

screen | light | dark
-- | -- | --
variable product with variations | ![Simulator Screen Shot - iPhone 11 - 2020-07-15 at 14 02 29](https://user-images.githubusercontent.com/1945542/87510124-929eb700-c6a5-11ea-8a9a-4c018795cc5c.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-15 at 14 03 43](https://user-images.githubusercontent.com/1945542/87510143-992d2e80-c6a5-11ea-97e9-bb926ac301a9.png)
variable product with variations - bottom sheet | ![Simulator Screen Shot - iPhone 11 - 2020-07-15 at 14 03 18](https://user-images.githubusercontent.com/1945542/87510141-98949800-c6a5-11ea-95f5-7ac6d7dd6974.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-15 at 14 03 46](https://user-images.githubusercontent.com/1945542/87510147-9a5e5b80-c6a5-11ea-95df-b6cb55819746.png)
variable product without variations | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 10 49 17](https://user-images.githubusercontent.com/1945542/87620864-32b21a00-c752-11ea-8062-0bfb93e95d77.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 10 48 39](https://user-images.githubusercontent.com/1945542/87620855-304fc000-c752-11ea-93f9-49a2e67d30dd.png)
variable product without variations - bottom sheet | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 10 49 18](https://user-images.githubusercontent.com/1945542/87620807-144c1e80-c752-11ea-9e63-c1c724cb21b2.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-16 at 10 48 58](https://user-images.githubusercontent.com/1945542/87620805-11e9c480-c752-11ea-806f-f2b868e1e90a.png)









Update release notes:
